src/relay/transforms/gradient_cell.cc

```diff
@@ -0,0 +1,320 @@
+    return wrapExpr(GetRef<TupleGetItem>(op), t);
+  }
+ private:
+  IRModule module_;
+};
+
+/*!
+* \brief Visitor to unwrap output
+*/
+class OutputVisitor: public ExprFunctor<Expr(const Expr&, const Type&)> {
+ public:
+  explicit OutputVisitor(IRModule module): module_(module) {}
+
+  Expr unwrapExpr(const Expr expr, const Type& type) {
+    if (auto* type_call = type.as<TypeCallNode>()) {
+      if (type_call->func.same_as(module_->GetGlobalTypeVar("GradCell"))) {
+        return CallNode::make(module_->GetGlobalVar("FromGradCell"), {expr});
+      }
+      return expr;
+    } else if (auto* type_anno = type.as<TupleTypeNode>()) {
+      tvm::Array<Expr> fields;
+      for (size_t i = 0; i < type_anno->fields.size(); i++) {
+        const Type& t = type_anno->fields[i];
+        fields.push_back(this->VisitExpr(TupleGetItemNode::make(expr, i), t));
+      }
+      Expr tuple = TupleNode::make(fields);
+      return tuple;
+    }
+
+    return expr;
+  }
+
+  Expr VisitExpr_(const CallNode* op, const Type& t) final {
+    return unwrapExpr(GetRef<Call>(op), t);
+  }
+
+  Expr VisitExpr_(const TupleGetItemNode* op, const Type& t) final {
+    return unwrapExpr(GetRef<TupleGetItem>(op), t);
+  }
+ private:
+  IRModule module_;
+};
+
+class GradientCellTransform: public ExprMutator, public TypeMutator {
+ public:
+  explicit GradientCellTransform(IRModule module):
+    module_(module) {
+      module_->ImportFromStd("gradient.rly");
+    }
+
+  /*!
+  * \brief apply GradientCell transformation and wrap function
+  * so that function type stays the same
+  * 
+  * input/output types should only be a combination of TupleTypes and TensorTypes
+  */
+  Expr transform(const Expr& e) {
+    auto* f = (e).as<FunctionNode>();
+    auto* transformed = this->Mutate(e).as<FunctionNode>();
+
+    if (e.same_as(GetRef<Function>(transformed))) {
+      return GetRef<Function>(transformed);
+    }
+
+    // wrap inputs of Tensor type using InputVisitor class
+    tvm::Array<Expr> args;
+    for (Var var : f->params) {
+      Expr wrappedInput = InputVisitor(module_).VisitExpr(var, var->checked_type());
+      args.push_back(wrappedInput);
+    }
+    Expr transformedExpr = CallNode::make(GetRef<Function>(transformed), args);
+
+    // unwrap outputs of GradCell type into Tensor type using OutputVisitor class
+    Expr tensorOutput = OutputVisitor(module_).VisitExpr(transformedExpr, transformed->ret_type);
+    return Function(f->params, tensorOutput, f->ret_type, Array<TypeVar>());
+  }
+
+  Expr VisitExpr_(const ConstantNode* op) final {
+    return CallNode::make(getGradCellConstructor(module_, "Raw"),
+                          {GetRef<Constant>(op)}, Attrs(), {op->checked_type()});
+  }
+
+  Expr VisitExpr_(const CallNode* call_node) final {
+    // optimize operators
+    if (auto* op = (call_node->op).as<OpNode>()) {
+      if (op->name.compare("add") == 0 && call_node->args.size() == 2 &&
```

Comment:
dont compare with name, and use op::get, with ==.
https://github.com/apache/incubator-tvm/blob/master/src/relay/backend/interpreter.cc#L431

Reply:
done thanks for review!